File: c/src/value_state.c
#include "value_state.h"
#include <stdlib.h>
#include <string.h>

void harmonic_escalation_branches_clear(HarmonicEscalationBranches* out) {
    if (!out) return;
    out->count = 0;
}

void harmonic_escalation_branches_set_more_harmonic_color(HarmonicEscalationBranches* out) {
    if (!out) return;
    out->branches[0] = HARMONIC_BRANCH_SEVENTHS;
    out->branches[1] = HARMONIC_BRANCH_ONE_INVERSION;
    out->count = 2;
}

/* Rule: Never more than one added harmonic degree per branch. */
int harmonic_escalation_branch_added_degrees(HarmonicEscalationBranch branch) {
    switch (branch) {
        case HARMONIC_BRANCH_SEVENTHS:
        case HARMONIC_BRANCH_ONE_INVERSION:
            return 1;  /* Each valid branch = exactly one added degree. */
        default:
            return -1; /* Invalid/unknown branch. */
    }
}

/* Invariant: at most one choice per branch; never more than one added degree per branch. */
bool harmonic_escalation_satisfies_invariant(const HarmonicEscalationBranches* b) {
    if (!b) return true;
    if (b->count > HARMONIC_ESCALATION_MAX_BRANCHES) return false;
    for (size_t i = 0; i < b->count; i++) {
        int degrees = harmonic_escalation_branch_added_degrees(b->branches[i]);
        if (degrees < 0 || degrees > HARMONIC_ESCALATION_MAX_ADDED_DEGREES_PER_BRANCH)
            return false;
    }
    return true;
}

void rhythmic_escalation_branches_clear(RhythmicEscalationBranches* out) {
    if (!out) return;
    out->count = 0;
}

void rhythmic_escalation_branches_set_more_groove(RhythmicEscalationBranches* out) {
    if (!out) return;
    out->branches[0] = RHYTHMIC_BRANCH_FORWARD_PUSH;
    out->branches[1] = RHYTHMIC_BRANCH_BACKWARD_LAYBACK;
    out->branches[2] = RHYTHMIC_BRANCH_NEUTRAL;
    out->count = 3;
}

/* Invariant: timing only. No branch may be "busier". */
bool rhythmic_escalation_satisfies_invariant(const RhythmicEscalationBranches* b) {
    if (!b) return true;
    if (b->count > RHYTHMIC_ESCALATION_MAX_BRANCHES) return false;
    return true;
}

ValueState value_state_present(double value) {
    ValueState vs = {VALUE_STATE_PRESENT, value, {0}};
    return vs;
}

ValueState value_state_missing(const char* reason) {
    ValueState vs = {VALUE_STATE_MISSING, 0.0, {0}};
    if (reason) {
        strncpy(vs.reason, reason, sizeof(vs.reason) - 1);
        vs.reason[sizeof(vs.reason) - 1] = '\0';
    }
    return vs;
}

ValueState value_state_invalid(const char* reason) {
    ValueState vs = {VALUE_STATE_INVALID, 0.0, {0}};
    if (reason) {
        strncpy(vs.reason, reason, sizeof(vs.reason) - 1);
        vs.reason[sizeof(vs.reason) - 1] = '\0';
    }
    return vs;
}

ValueStateSet value_state_set_create(void) {
    ValueStateSet set = {0};
    for (int i = 0; i < VALUE_TYPE_COUNT; i++) {
        set.states[i] = value_state_missing("not_derived");
    }
    return set;
}

void value_state_set_set(ValueStateSet* set, ValueType type, ValueState state) {
    if (!set || type >= VALUE_TYPE_COUNT) return;
    set->states[type] = state;
}

ValueState value_state_set_get(const ValueStateSet* set, ValueType type) {
    if (!set || type >= VALUE_TYPE_COUNT) return value_state_missing("invalid_type");
    return set->states[type];
}

void value_state_set_preserve_absence(ValueStateSet* dest, const ValueStateSet* src) {
    if (!dest || !src) return;
    for (int i = 0; i < VALUE_TYPE_COUNT; i++) {
        if (value_state_is_missing(&src->states[i]) || value_state_is_invalid(&src->states[i])) {
            dest->states[i] = src->states[i];
        }
    }
}

bool value_state_is_present(const ValueState* vs) {
    return vs && vs->kind == VALUE_STATE_PRESENT;
}

bool value_state_is_missing(const ValueState* vs) {
    return vs && vs->kind == VALUE_STATE_MISSING;
}

bool value_state_is_invalid(const ValueState* vs) {
    return vs && vs->kind == VALUE_STATE_INVALID;
}

double value_state_get_value(const ValueState* vs) {
    return vs && vs->kind == VALUE_STATE_PRESENT ? vs->value : 0.0;
}

const char* value_state_get_reason(const ValueState* vs) {
    return vs ? vs->reason : "";
}

ValueState value_state_derive(ValueType type, double intent_magnitude, bool constrained) {
    (void)type;
    if (constrained) return value_state_missing("constrained");
    if (intent_magnitude < 0.0 || intent_magnitude > 1.0)
        return value_state_invalid("magnitude_out_of_range");
    return value_state_present(intent_magnitude);
}

bool value_state_validate(const ValueState* vs, ValueType type) {
    (void)type;
    if (!vs) return false;
    if (vs->kind == VALUE_STATE_INVALID) return false;
    if (vs->kind == VALUE_STATE_PRESENT && (vs->value < 0.0 || vs->value > 1.0)) return false;
    return true;
}

File: rust/src/value_state.rs
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ValueStateKind {
    Present,
    Missing,
    Invalid,
}

#[derive(Debug, Clone)]
pub enum ValueState {
    Present(f64),
    Missing { reason: String },
    Invalid { reason: String },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ValueType {
    TempoBias,
    HarmonicTension,
    RhythmicDensity,
    DynamicRange,
}

/// HARMONIC ESCALATION RULE (codified):
///
/// One branch may choose:
///   - triad + inversion
///   OR
///   - triad + 7th
///
/// Never both.
/// Never more than one added harmonic degree.
///
/// Forbidden: 7ths+inversions, multiple extensions (9th, 11th), cascading "well since we can..."
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum HarmonicEscalationBranch {
    /// triad + 7th. One added degree.
    Sevenths,
    /// triad + inversion. One added degree.
    OneInversion,
}

pub const HARMONIC_ESCALATION_MAX_ADDED_DEGREES_PER_BRANCH: usize = 1;

/// Mutually exclusive branches; pick one. count is 0, 1, or 2.
#[derive(Debug, Clone)]
pub struct HarmonicEscalationBranches {
    pub branches: Vec<HarmonicEscalationBranch>,
}

impl HarmonicEscalationBranches {
    pub const MAX_BRANCHES: usize = 2;

    pub fn empty() -> Self {
        Self {
            branches: Vec::new(),
        }
    }

    pub fn more_harmonic_color() -> Self {
        Self {
            branches: vec![
                HarmonicEscalationBranch::Sevenths,
                HarmonicEscalationBranch::OneInversion,
            ],
        }
    }

    /// Invariant: at most one choice per branch; never more than one added degree per branch.
    pub fn satisfies_invariant(&self) -> bool {
        if self.branches.len() > Self::MAX_BRANCHES {
            return false;
        }
        for branch in &self.branches {
            if HarmonicEscalationBranch::added_degrees(*branch) > HARMONIC_ESCALATION_MAX_ADDED_DEGREES_PER_BRANCH
            {
                return false;
            }
        }
        true
    }
}

impl HarmonicEscalationBranch {
    /// Returns added harmonic degrees for a branch. Must be <= MAX_ADDED_DEGREES_PER_BRANCH (1).
    pub fn added_degrees(self) -> usize {
        match self {
            HarmonicEscalationBranch::Sevenths | HarmonicEscalationBranch::OneInversion => 1,
        }
    }
}

/// RHYTHMIC ESCALATION RULE (codified):
///
/// MICRO_TIMING_ONLY: push/pull, swing bias, humanization window.
///
/// Does NOT: add notes, change pattern, increase density, add fills,
/// change velocity curves beyond a small window.
///
/// Branches may differ by: forward push, backward layback, neutral.
/// No branch may be "busier".
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RhythmicEscalationBranch {
    /// timing only: push forward
    ForwardPush,
    /// timing only: lay back
    BackwardLayback,
    /// timing only: neutral
    Neutral,
}

#[derive(Debug, Clone)]
pub struct RhythmicEscalationBranches {
    pub branches: Vec<RhythmicEscalationBranch>,
}

impl RhythmicEscalationBranches {
    pub const MAX_BRANCHES: usize = 3;

    pub fn empty() -> Self {
        Self { branches: Vec::new() }
    }

    pub fn more_groove() -> Self {
        Self {
            branches: vec![
                RhythmicEscalationBranch::ForwardPush,
                RhythmicEscalationBranch::BackwardLayback,
                RhythmicEscalationBranch::Neutral,
            ],
        }
    }

    /// Invariant: timing only. No branch may be "busier".
    pub fn satisfies_invariant(&self) -> bool {
        self.branches.len() <= Self::MAX_BRANCHES
    }
}

impl ValueState {
    pub fn present(value: f64) -> Self {
        ValueState::Present(value)
    }

    pub fn missing(reason: impl Into<String>) -> Self {
        ValueState::Missing {
            reason: reason.into(),
        }
    }

    pub fn invalid(reason: impl Into<String>) -> Self {
        ValueState::Invalid {
            reason: reason.into(),
        }
    }

    pub fn is_present(&self) -> bool {
        matches!(self, ValueState::Present(_))
    }

    pub fn is_missing(&self) -> bool {
        matches!(self, ValueState::Missing { .. })
    }

    pub fn is_invalid(&self) -> bool {
        matches!(self, ValueState::Invalid { .. })
    }

    pub fn get_value(&self) -> f64 {
        match self {
            ValueState::Present(v) => *v,
            _ => 0.0,
        }
    }

    pub fn get_reason(&self) -> &str {
        match self {
            ValueState::Missing { reason } | ValueState::Invalid { reason } => reason,
            _ => "",
        }
    }

    pub fn derive(_value_type: ValueType, intent_magnitude: f64, constrained: bool) -> Self {
        if constrained {
            ValueState::missing("constrained")
        } else if intent_magnitude < 0.0 || intent_magnitude > 1.0 {
            ValueState::invalid("magnitude_out_of_range")
        } else {
            ValueState::present(intent_magnitude)
        }
    }

    pub fn validate(&self, _value_type: ValueType) -> bool {
        match self {
            ValueState::Invalid { .. } => false,
            ValueState::Present(v) => *v >= 0.0 && *v <= 1.0,
            ValueState::Missing { .. } => true,
        }
    }
}

#[derive(Debug, Clone)]
pub struct ValueStateSet {
    states: HashMap<ValueType, ValueState>,
}

impl ValueStateSet {
    pub fn new() -> Self {
        let mut states = HashMap::new();
        for vt in [
            ValueType::TempoBias,
            ValueType::HarmonicTension,
            ValueType::RhythmicDensity,
            ValueType::DynamicRange,
        ] {
            states.insert(vt, ValueState::missing("not_derived"));
        }
        Self { states }
    }

    pub fn set(&mut self, value_type: ValueType, state: ValueState) {
        self.states.insert(value_type, state);
    }

    pub fn get(&self, value_type: ValueType) -> ValueState {
        self.states
            .get(&value_type)
            .cloned()
            .unwrap_or_else(|| ValueState::missing("invalid_type"))
    }

    pub fn preserve_absence(&mut self, other: &ValueStateSet) {
        for (vt, state) in &other.states {
            if state.is_missing() || state.is_invalid() {
                self.states.insert(*vt, state.clone());
            }
        }
    }
}

impl Default for ValueStateSet {
    fn default() -> Self {
        Self::new()
    }
}
